cachi2/core/package_managers/gomod.py

```diff
@@ -537,29 +598,63 @@ def go_list_deps(pattern: Literal["./...", "all"]) -> Iterator[ParsedPackage]:
 
     _validate_local_replacements(all_modules, app_dir)
 
-    return main_module, all_modules, all_packages
+    return ResolvedGoModule(main_module, all_modules, all_packages, modules_in_go_sum)
+
+
+def _parse_go_sum(module_dir: RootedPath) -> frozenset[ModuleID]:
+    """Return the set of modules present in the go.sum file in the specified directory.
+
+    A module is considered present if the checksum for its .zip file is present. The go.mod file
+    checksums are not relevant for our purposes.
+    """
+    go_sum = module_dir.join_within_root("go.sum")
+    if not go_sum.path.exists():
+        return frozenset()
+
+    modules: list[ModuleID] = []
+
+    # https://github.com/golang/go/blob/d5c5808534f0ad97333b1fd5fff81998f44986fe/src/cmd/go/internal/modfetch/fetch.go#L507-L534
+    lines = go_sum.path.read_text().splitlines()
+    for i, go_sum_line in enumerate(lines):
+        parts = go_sum_line.split()
+        if not parts:
+            continue
+        if len(parts) != 3:
+            # - when readGoSum (see link above) encounters a malformed line, it returns an error
+            # - every call to readGoSum ignores that error
+            #   - https://github.com/golang/go/blob/d5c5808534f0ad97333b1fd5fff81998f44986fe/src/cmd/go/internal/modfetch/fetch.go#L495
+            #   - https://github.com/golang/go/blob/d5c5808534f0ad97333b1fd5fff81998f44986fe/src/cmd/go/internal/modfetch/fetch.go#L836
+            # - readGoSum modifies a global variable (indirectly, it's passed in as a param)
+            # - => readGoSum will parse the non-broken part of go.sum and then silently fail
+            # - replicate the bug here, because it means that go only uses the non-broken part
```

Comment:
Ouch! :harold:

I agree that we should have the same behavior as `go`, but maybe this should be something we also report to the user somehow? I feel like this is the type of error that would go unnoticed for the longest of time (and maybe the warning logs here would be unnoticed as well).

Not really sure how to report it... probably a matter for a separate PR, though.

Reply:
The only better way to report it would be to raise an exception IMO :slightly_smiling_face:

Reply:
And now that I'm testing things locally, it looks like missing checksums cause `go build` to fail anyway, so reporting missing checksums is probably a bit redundant anyway :sweat_smile: